### PR TITLE
KeyBuilder: fix hashing of dataclasses/attrs, add test

### PR DESCRIPTION
**Pull request by matthiasdiener**
The test fails without this PR. Followup of #147.

_Please squash_

### CODE REVIEW

File: pytools/test/test_persistent_dict.py [hunk 1]
```diff
@@ -518,6 +518,51 @@ class TagClass2(Tag):
...
+
+    # Class types must be encoded in hash
+    assert keyb(MyDC2("hi", 1)) != keyb(MyDC("hi", 1))
+
+    try:
+        import attrs
+    except ImportError:
+        from warnings import warn
+        warn("attrs not installed, skipping rest of the test")
+        return
```
Comment (matthiasdiener) [inline pytools/test/test_persistent_dict.py:545]: Not sure what's better here, `warn` or `pytest.importorskip`.

Comment (inducer) [reply pytools/test/test_persistent_dict.py:545]: Break the `attrs` bit out into a separate test and use `importorskip`.

**Review by inducer — APPROVED** [review]
LGTM other than the one mini wrinkle.

File: pytools/test/test_persistent_dict.py (hunk 1 above)
Comment (matthiasdiener) [reply pytools/test/test_persistent_dict.py]: Done in 4a1ac64

**Pull request merged**